### PR TITLE
Complex changes

### PR DESCRIPTION
**Pull request by Hadatko**
-- different approach to serialize data and serialize also repeated data.
 -- generate allowed numbers
-- generate masks and offsets

### CODE REVIEW

File: lang/cpp/src/Ubx.hpp [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+#ifndef __UBX_H__
+#define __UBX_H__
+
+#include <cstdint>
+#include <string>
+
+#define UBX_CHANGE_VAR_BIT(name, val, map, offset) (name = (name & ~map) | ((val << offset) & map))
+#define UBX_GET_VAR_BIT(name, map, offset) ((name & map) >> offset)
+
+namespace ubx
```
Comment (qcabrol) [inline lang/cpp/src/Ubx.hpp:12]: missing comment headers

File: lang/cpp/src/Ubx.cpp [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+        uint16_t index = 0;
+
+        // serialize data
+        writeByte(SerializeCommon::ubxHead0, buf, index);
+        writeByte(SerializeCommon::ubxHead1, buf, index);
+        writeByte(getClassId(), buf, index);
+        ckSum.update(getClassId());
+        writeByte(getMessageId(), buf, index);
+        ckSum.update(getMessageId());
+        writeByte(getDataSize() & 0x00ff, buf, index);
```
Comment (qcabrol) [inline lang/cpp/src/Ubx.cpp:44]: little endian order with lsb added first.

File: lang/cpp/src/Ubx.cpp [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+        ckSum.update(getDataSize() & 0x00ff);
+        writeByte(getDataSize() >> 8, buf, index);
+        ckSum.update(getDataSize() >> 8);
+        for (uint16_t i = 0; i < getDataSize(); i++)
+        {
+            writeByte(getDataStartAddress()[i], buf, index);
```
Comment (qcabrol) [inline lang/cpp/src/Ubx.cpp:50]: how are we sure of the bytes oendianness order in the buffer ? This is handled lower ?

File: lang/cpp/src/parseUbxMessage.hpp [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+#include "messages/LOG.hpp"
+#include "messages/MGA.hpp"
+#include "messages/MON.hpp"
+#include "messages/NAV.hpp"
+#include "messages/TEST.hpp"
+#include "messages/TIM.hpp"
+
+namespace ubx
+{
+    class ParseUbxMessage : public ParseUbxMessageBase
```
Comment (qcabrol) [inline lang/cpp/src/parseUbxMessage.hpp:25]: comment headeer

File: lang/cpp/src/parseUbxMessageBase.cpp [hunk 5]
```diff
@@ -0,0 +1,75 @@
+#include "parseUbxMessageBase.hpp"
+#include <cstring>
+
+using namespace ubx;
+using namespace std;
+
+int32_t ParseUbxMessageBase::findUbxMessageStartIdx(const uint8_t *buf, uint16_t len) const
```
Comment (qcabrol) [inline lang/cpp/src/parseUbxMessageBase.cpp:7]: prototypes with header documentation missing

File: ubx/Types.py [hunk 6]
```diff
@@ -10,13 +10,25 @@
 # Note: All classes listed here must be also listed in UBXMessages.py,
 # function classFromMessageClass !!
 
+def parseBitfields(bitfield):
+    for b in bitfield:
+        if type(bitfield[b]) == dict:
+            s = bitfield[b]["s"]
```
Comment (qcabrol) [inline ubx/Types.py]: better va r names ?

File: ubx/Types.py [hunk 7]
```diff
@@ -145,13 +158,15 @@ def serialize(self, val):
             raise Exception(err)
 
 class U:
-    """Variable-length array of unsigned chars."""
+    """Variable-length array of unsigned chars.."""
```
Comment (qcabrol) [inline ubx/Types.py]: stray point change

File: ubx/Types.py (hunk 6 above)
Comment (qcabrol) [reply ubx/Types.py]: to mention one is the bitshift and the other a byteshift ? That or a minor comment

File: ubx/generateCPP.py [hunk 8]
```diff
@@ -52,166 +52,456 @@ def makeMemberDecl(typ, name):
         return "{} {};".format(typ, name)
 
 
-def makeMessageStruct(file, className, Message, indent=4):
-    messageName = getMessageName(Message, withUBX=False, withMessageClass=False)
-    if messageName == className:
-        messageName = messageName + "_"
-    fullClassName = "{}::{}".format(className, messageName)
+def getBitName(field,  bit):
```
Comment (qcabrol) [inline ubx/generateCPP.py:58]: small comment doc for each routine

File: ubx/generateCPP.py [hunk 9]
```diff
@@ -52,166 +52,456 @@ def makeMemberDecl(typ, name):
...
+    return field[0] + "_" + bit
+
+
+def getBitNameUp(field, bit):
```
Comment (qcabrol) [inline ubx/generateCPP.py]: getBitNameUpperCase() 
would be more transparent as a name

File: ubx/generateCPP.py [hunk 10]
```diff
@@ -52,166 +52,456 @@ def makeMemberDecl(typ, name):
...
+def generateEnumDefinitions(fields, file, repeated=False):
+    if fields:
+        if repeated:
+            name = "repeated_"
+        else:
+            name = ""
+        for field in fields:
+            allowed = getattr(field[1], "allowed", {})
+            if len(allowed.keys()) > 0:
+                file.write(f"        enum _{name}{field[0]}_allowed{{\n")
```
Comment (qcabrol) [inline ubx/generateCPP.py:81]: why the spaces after "  f"  " ? Identation purpose ?

File: ubx/generateCPP.py [hunk 11]
```diff
@@ -52,166 +52,456 @@ def makeMemberDecl(typ, name):
...
+def getBitName(field,  bit):
+    return field[0] + "_" + bit
+
+
+def getBitNameUp(field, bit):
+    return getBitName(field, bit).upper()
+
+
+def generateEnumDefinitions(fields, file, repeated=False):
+    if fields:
```
Comment (qcabrol) [inline ubx/generateCPP.py:73]: missing comment for the routine

Comment (qcabrol) [reply ubx/generateCPP.py:73]: Generate bitfield allowed values as C enums from the corresponding py classes allowed dict fields.

File: ubx/generateCPP.py [hunk 12]
```diff
@@ -52,166 +52,456 @@ def makeMemberDecl(typ, name):
...
+            name = "repeated_"
+        else:
+            name = ""
+        for field in fields:
+            allowed = getattr(field[1], "allowed", {})
+            if len(allowed.keys()) > 0:
+                file.write(f"        enum _{name}{field[0]}_allowed{{\n")
+                keys = {}
+                for a in allowed:
+                    key = allowed[a].replace(' ', '_').replace('-', '_')
```
Comment (qcabrol) [inline ubx/generateCPP.py:84]: Should we conside other special separators like "." here ? (apparently none of the allowed fields use the . at the moment)

File: ubx/generateCPP.py [hunk 13]
```diff
@@ -1,5 +1,5 @@
 #!/usr/bin/env python3
-"""Generate CPP code from the UBX message defintions in UBX/."""
+"""Generate CPP code from the UBX message definitions in UBX/."""
```
Comment (qcabrol) [inline ubx/generateCPP.py:2]: missing readme for the updated gen script ?

File: ubx/generateCPP.py [hunk 14]
```diff
@@ -52,166 +52,456 @@ def makeMemberDecl(typ, name):
...
+                file.write(f"\n")
+                for bit in bitfield:
+                    file.write(
+                        f"void ubx::{className}::_data::{name2}{getBitName(field, bit)}(uint8_t val{nDefCom}{nDef}){{ UBX_CHANGE_VAR_BIT({repeatedName}{field[0]}, val, {name}_{getBitNameUp(field, bit)}_MASK, {name}_{getBitNameUp(field, bit)}_OFFSET); }}\n")
+                    file.write(
+                        f"uint8_t ubx::{className}::_data::{name2}{getBitName(field, bit)}({nDef})const{{ return UBX_GET_VAR_BIT({repeatedName}{field[0]}, {name}_{getBitNameUp(field, bit)}_MASK, {name}_{getBitNameUp(field, bit)}_OFFSET); }}\n")
+
+
+def makeMessageStruct(file, className, Message):
+    messageName = getMessageName(
```
Comment (qcabrol) [inline ubx/generateCPP.py:165]: where is getMessageName() defined ?

File: ubx/generateCPP.py [hunk 15]
```diff
@@ -52,166 +52,456 @@ def makeMemberDecl(typ, name):
...
+        Message, withUBX=False, withMessageClass=False)
+    # fullClassName = "{}::{}".format(className, messageName)
+    fullClassNameUnder = "{}_{}".format(className, messageName)
+    fields = getClassMembers(Message.Fields)
```
Comment (qcabrol) [inline ubx/generateCPP.py:169]: where is  getClassMembers() defined ?

Comment (qcabrol) [reply ubx/generateCPP.py:169]: ubx/introspect.py:51:def getClassMembers(cls, predicate=None):

File: ubx/generateCPP.py (hunk 14 above)
Comment (qcabrol) [reply ubx/generateCPP.py:165]: nevermind, found on introspect.py

File: ubx/generateCPP.py [hunk 16]
```diff
@@ -52,166 +52,456 @@ def makeMemberDecl(typ, name):
...
-        shortName = fileName.relative_to(topdir) # Starts with lang/cpp...
-        # fileName = "lang/cpp/src/messages/{}.h".format(className)
-        file = open(fileName, 'w')
+        fileName = cppsrcdir.joinpath("messages", className+".hpp")
+        shortName = fileName.relative_to(topdir)  # Starts with lang/cpp...
+        file = open(fileName, 'w+')
         file.write("// File {}\n".format(shortName))
         file.write("// Auto-generated by pyUBX {} v{} on {}\n".format(
             os.path.basename(__file__),
             __version__,
```
Comment (qcabrol) [inline ubx/generateCPP.py:364]: do we want to update the version number as you did major changes ?

**Review by qcabrol — CHANGES_REQUESTED** [review]

File: ubx/generateCPP.py (hunk 10 above)
Comment (Hadatko) [reply ubx/generateCPP.py:81]: yes

File: ubx/generateCPP.py (hunk 12 above)
Comment (Hadatko) [reply ubx/generateCPP.py:84]: Can be added once needed, but currently they are not.

File: ubx/generateCPP.py (hunk 16 above)
Comment (Hadatko) [reply ubx/generateCPP.py:364]: yes we can

File: lang/cpp/src/Ubx.cpp (hunk 3 above)
Comment (Hadatko) [reply lang/cpp/src/Ubx.cpp:50]: No it was prototyped as LE

**Review by qcabrol — APPROVED** [review]
Changes on documentation and fix on packet size calculation have landed. Happy with the current PR I think we can move forward

**Pull request merged**